### DISCUSSION TITLE
[Feature] Allow setting the title from the renamed file

### DISCUSSION THREAD

**Discussion by sebastiang-ro** [jonaswinkler/paperless-ng#815] [Feature Requests]
Currently the initial title is set from the original filename. In my case this comes from my scanner and is rather meaningless (e.g. duplex27-02-2021-111449.pdf for a duplex scan from today).

I have started to import a lot of files and the matching for correspondent, document type and tags is pretty good. However the title is a problem for me, as I don't want to rename hundreds or thousands of titles. I have enabled file name renaming ({correspondent}/{created_year}/{document_type}-{created}) and it would be good if I could somehow automatically change the title to the renamed file name.

**Comment by jonaswinkler** [comment]
I'm not sure about this.

* This would simply copy data that's available in other fields into the title. No added benefit for searching purposes.
* What if you use `{title}` in the filename format along with other placeholders?
* While I agree that filenames from documents produced by scanners are rarely useful, some other filenames are, and I'm not sure how I'd decide whether to overwrite the given title with something constructed from metadata.
* What about saving updated correspondents / types? should the title update as well? (What if you manually changed the title?)

**Reply by jgillula** [reply]
Just my two cents:

- I agree there's no benefit for searching, but for me the benefit is the ability to see at a glance the fields that I've designated as important combined into one field.
- To prevent recursive use of ```{title}```, I think you would need a separate config setting for *just* the title, which supports all the other fields *except* ```{title}```. For example, ```PAPERLESS_TITLE_FORMAT```.
- Choosing whether source filenames are important or not seems like a decision we could leave to the user. For me (as with OP) source filenames are *never* useful. If we make this an option where the default is to preserve the original filename, then users could choose what fits their use cases.
- I can see three ways to do this, in order of increasing complexity:
   - Automagic title creation happens only when a document is first processed, and after that titles can be manually edited, and updating fields does not automatically update the title, or
   - Titles are set at creation, can be manually edited, and updating any field automatically updates the title, overwriting any manual edits,
   - Titles are set at creation, cannot be manually edited, and updating any field automatically updates the title, (manual editing is only allowed if automagic title creation is disabled),
   - Titles are set at creation, can be manually edited, and manual edits set a per-document flag, which prevents further automatic title updates; otherwise titles are updated any time a field is updated

Of those options, I'd be fine with just the first option. It also might be worth adding another field for the original filename, in case some users find those useful but also want custom titles.

**Comment by sebastiang-ro** [comment]
I agree, it doesn't make sense in an automatic kind of way or under specific circumstances.

Maybe it would make more sense to have a bulk "Title" assign function in the "Document view", where I can select a specific subset of documents (through filtering for tags, types). With an additional dropdown I could then assign a plain string or a string with placeholders (for existing data fields like {document_type} except the previous {title}) as new title for all selected documents.

Just as an example: 100 documents where the matcher has already successfully matched the document type "Bank statements" and correspondent "My Bank" as well as the "Date created". But all titles are "scan-yyyy-mm-dd-xyz" because of the file name of the scanned file. With the new functionality I could now easily set the title for all documents to: "{document-type} - {month}/{year}"

**Reply by jce-zz** [reply]
I would absolutely love this function. I have a large collection of electronic files which I would like to import into paperless and I would love to set the title to something meaningful in a bulk fashion. The file name is normally not really good enough.

**Comment by MichaelHochriegl** [comment]
I too would really like the ability to set the title by a regex. For example I want to have a proper title for my paychecks corresponding to the month they are for.

Don't see how I could achieve this with a tag, as I would have to create a tag for every "year-month" combination, which will get messy pretty quickly.

**Comment by MAngel666** [comment]
+1

**Comment by jce-zz** [comment]
I would also think such a function would be really helpful, especially when importing/consuming a lot of similar documents/files.

**Comment by jgillula** [comment]
For anyone looking for a solution, I wrote [`paperless-ngx-postprocessor`](https://github.com/jgillula/paperless-ngx-postprocessor). It allows you to automatically set the title using the usual document metadata. The only catch is that it only works with [paperless-ngx](https://github.com/paperless-ngx/paperless-ngx).

**Comment by jce-zz** [comment]
Wow! This sounds brilliant! I need to give it a try asap. From reading the description, it really seems to do mostly everything I ever imagined. Ives’s only wondering, if you plan to also add support for multiple criteria on the command line?

**Reply by jgillula** [reply]
I wasn't planning on it since I don't have much need for it myself (and I have so many other projects I'm behind on), but I'm open to accepting a PR if you wanted to.
